[mlir][c] Expose AsmState.

jpienaar opened the pull request:
Enable usage where capturing AsmState is good. Haven't plumbed through to python yet. This also only changes one C API to verify plumbing.

makslevental commented on mlir/include/mlir-c/IR.h:
```diff
@@ -383,6 +384,23 @@ mlirOperationStateAddAttributes(MlirOperationState *state, intptr_t n,
+// functions to maximize the possibility of compatibility over time.
+//===----------------------------------------------------------------------===//
+
+/// Creates new AsmState, as with AsmState the IR should not be mutated
```
Nit: should this say maybe "as with OperationState"? Not sure.

jpienaar replied:
Not sure I follow, OperationState is created while mutating the IR.

makslevental commented on mlir/include/mlir-c/IR.h:
```diff
@@ -819,6 +837,11 @@ MLIR_CAPI_EXPORTED void mlirValuePrintAsOperand(MlirValue value,
                                                 void *userData);
 
+/// Prints a value as an operand (i.e., the ValueID) using prepopulated state.
+MLIR_CAPI_EXPORTED void
```
Is there a way to unify this with `mlirValuePrintAsOperand`? Maybe awkward.

jpienaar replied:
Yeah I wish overloading was a thing. Problem is as AsmState is initialized with OpPrintingFlags,  if you pass in both is confusing ... that being said AsmState is the better one to pass in, currently its being created in opaque way. Tagged union seems the best I could do here if I wanted to unify.

jpienaar replied:
I went the other way: I made this the call signature and made it easier to construct the original using building blocks.

AsmState usage here subsumes OpPrintingFlags one and allows for much more efficient printing (avoiding the quadratic behavior if one walks & prints and ends up recreating the state over and over). Also makes it much clearer that there is a cost rather then hiding it. I had to add one helper to do so and changed the corresponding Python interface. I think probably all of the other print ones should be similarly changed, but wanted to check this change first.

makslevental reviewed (approved):
LGTM modulo small nits/questions.

jpienaar commented on mlir/test/python/ir/value.py:
Ug, seems formatter got creative. I'll revert this file.

makslevental replied:
Lol G style guide strikes again

jpienaar committed:
[mlir][c] Expose AsmState.

Enable usage where capturing AsmState is good. Haven't plumbed through to python yet.
